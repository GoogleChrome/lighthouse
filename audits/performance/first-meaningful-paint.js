/**
 * @license
 * Copyright 2016 Google Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

'use strict';

const FMPMetric = require('../../metrics/performance/first-meaningful-paint');
const Audit = require('../audit');

class FirstMeaningfulPaint extends Audit {
  /**
   * @override
   */
  static get tags() {
    return ['Performance'];
  }

  /**
   * @override
   */
  static get name() {
    return 'first-meaningful-paint';
  }

  /**
   * @override
   */
  static get description() {
    return 'Has a good first meaningful paint';
  }

  /**
   * Audits the page to give a score for First Meaningful Paint.
   * @see  https://github.com/GoogleChrome/lighthouse/issues/26
   * @param {!Artifacts} artifacts The artifacts from the gather phase.
   * @return {!AuditResult} The score from the audit, ranging from 0-100.
   */
  static audit(artifacts) {
    return FMPMetric
        .parse(artifacts.traceContents)
        .then(fmp => {
          if (fmp.err) {
            return {
              score: -1
            };
          }

          // Roughly an exponential curve.
          // < 1000ms: penalty=0
          // 3000ms: penalty=90
          // >= 5000ms: penalty=100
          const power = (fmp.duration - 1000) * 0.001 * 0.5;
          const penalty = power > 0 ? Math.pow(10, power) : 0;
          let score = 100 - penalty;

          // Clamp the score to 0 <= x <= 100.
          score = Math.min(100, score);
          score = Math.max(0, score);

          return {
            duration: `${fmp.duration.toFixed(2)}ms`,
            score: Math.round(score)
          };
        }, _ => {
          // Recover from trace parsing failures.
          return {
            score: -1
          };
        })
        .then(result => {
          return FirstMeaningfulPaint.generateAuditResult(
              result.score, result.duration, undefined,
              this.generateRecommendedActions(result.score, artifacts)
          );
        });
  }

  /**
   * Generates some recommended actions for FMP.
   * @param {number} score The score to determine if we need to bother.
   * @param {!Artifacts} artifacts The artifacts from the gather phase.
   * @return {Array<!AuditRecommendedAction>}
   */
  static generateRecommendedActions(score, artifacts) {
    // If there is a score of 100 no further action is required.
    if (score === 100) {
      return undefined;
    }

    // Attempt to figure out what the blocking resources are in the page.
    const recommendedActions = artifacts.blockingResources
        .map(resource => {
          let suggestion = '';

          switch (resource.data._resourceType._name) {
            case 'script':
              suggestion = 'should have either async or defer attribute';
              break;

            case 'stylesheet':
              suggestion = 'should be moved inline to the HTML';
              break;

            default:
              suggestion = 'should be removed';
              break;
          }

          return {
            title: `${resource.data.url} ${suggestion}`,
            details: [
              `blocks for ${(resource.data.endTime - resource.data.startTime).toFixed(2)}ms`
            ]
          };
        });

    return recommendedActions;
  }
}

module.exports = FirstMeaningfulPaint;
