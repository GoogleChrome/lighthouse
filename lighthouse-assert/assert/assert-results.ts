import { ICollatedResult, IResult, IExpectation, ICollatedAudit } from './types';
import { DifferenceFactory } from './difference/diff-factory';

const FINAL_URL = 'final url';

export class AssertResults {
  private collatedResults: Array<ICollatedResult> = [];
  /**
   * Collate results on each expectation.
   */
  collate(results: Array<IResult>, expectations: Array<IExpectation>) {
    expectations.forEach(
      (expectation, index) => {
        this.collatedResults.push(this.collateAuditResults(results[index], expectation))});
    return this.collatedResults;
  }

  /**
   * Collate results into comparisons of actual and expected scores on each audit.
   * @param {{url: string, audits: IResult}} actual
   * @param {{url: string, audits: IResult}} expected
   * @return {{finalUrl: Object, audits: !Array<ICollatedAudit>}}
   */
  private collateAuditResults(actual: IResult, expected: IExpectation): ICollatedResult {
    const collatedAudits: Array<ICollatedAudit> = [];
    for (const auditName in expected.audits) {
      const actualResult = actual.audits[auditName];
      if (!actualResult) {
        throw new Error(`Config did not trigger run of expected audit ${auditName}`);
      }

      const expectedResult = expected.audits[auditName];
      const diff = DifferenceFactory.findDifference(auditName, actualResult, expectedResult);
      collatedAudits.push({
        category: auditName,
        actual: actualResult,
        expected: expectedResult,
        equal: !Object.keys(diff).length,
        diff: diff
      });
    }

    return {
      finalUrl: {
        category: FINAL_URL,
        actual: actual.url,
        expected: expected.url,
        equal: actual.url === expected.url
      },
      audits: collatedAudits
    };
  }
}
