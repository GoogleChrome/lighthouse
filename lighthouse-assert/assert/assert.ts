export interface Result {
  url: string;
  audits: ActualAudits;
}

export interface Expectation {
  url: string;
  audits: ExpectedAudits;
}

interface ActualAudits {
  [key: string]: ActualAudit
}

interface ActualAudit {
  [key: string]: any;
}

interface NoneObjectActualAudit {
  score: string|number;
  [key: string]: string|number;
}

interface BooleanActualAudit {
  score: boolean;
  [key: string]: boolean;
}

interface ExpectedAudits {
  [key: string]: ExpectedAudit;
}

interface ExpectedAudit {
  score: ExpectedScore;
  [key: string]: ExpectedScore;
}

interface NoneObjectExpectedAudit {
  score: ExpectedScore;
  [key: string]: ExpectedScore;
}

interface BooleanExpectedAudit {
  score: boolean;
  [key: string]: boolean;
}

interface ExpectedScore {
  error: string;
  warn: string;
  [key: string]: string;
}

interface NormalizedExpectedScore {
  error: number;
  warn: number;
}


interface CollatedResult {
  finalUrl: {
    category: string;
    actual: string;
    expected: string;
    equal: boolean;
  };
  audits: Array<CollatedAudit>;
}

interface CollatedAudit {
  category: string;
  actual: ActualAudit;
  expected: ExpectedAudit;
  equal: boolean;
  diff: Diff|Object;
}

export interface Diff extends Object {
  path?: string;
  actual?: string|number|boolean;
  expected?: ExpectedScore|boolean;
}

export interface StatusCounts extends Object {
  passed: number;
  failed: number;
}

const FINAL_URL = 'final url';

export class Assert {
  public collatedResults: Array<CollatedResult> = [];
  private results: Array<Result>;
  private expectations: Array<Expectation>;

  /**
   * Constructor
   * @param {Result} lhResults
   * @param {Array<Expectation>} expectations
   * @return {boolean}
   */
  constructor(lhResults: Array<Result>, expectations: Array<Expectation>) {
    this.results = lhResults;
    this.expectations = expectations;
  }

  /**
   * Verify if results are equal
   * @return {boolean}
   */
  equal(): boolean {
    const notCollatedResults = this.collatedResults.find(result => {
      const auditResults = result.audits.find(audit => !audit.equal);
      return (auditResults && !auditResults.equal) ? true : false;
    });
    return notCollatedResults ? false : true;
  }

  /**
   * Collate results on each expectation.
   */
  collate() {
    this.expectations.forEach((expectation, index) => {
      this.collatedResults.push(this.collateAuditResults(this.results[index], expectation))
    });
  }

  /**
   * Get status counts for collated results
   */
  getStatusCounts() {
    let statusCounts: StatusCounts = {
      passed: 0,
      failed: 0
    };

    for (const collatedResult of this.collatedResults) {
      // @todo include other results then audits
      for (const audit of collatedResult.audits) {
        (audit && audit.equal) ? statusCounts.passed += 1 : statusCounts.failed += 1;
      }
    }
    return statusCounts;
  }

  /**
   * Collate results into comparisons of actual and expected scores on each audit.
   * @param {{url: string, audits: Result}} actual
   * @param {{url: string, audits: Result}} expected
   * @return {{finalUrl: Object, audits: !Array<CollatedAudit>}}
   */
  private collateAuditResults(actual: Result, expected: Expectation): CollatedResult {
    const auditNames = Object.keys(expected.audits);
    const collatedAudits: Array<CollatedAudit> = [];
    auditNames.forEach(auditName => {
      const actualResult = actual.audits[auditName];
      if (!actualResult) {
        throw new Error(`Config did not trigger run of expected audit ${auditName}`);
      }

      const expectedResult = expected.audits[auditName];
      const diff = DifferenceFactory.findDifference(auditName, actualResult, expectedResult);
      collatedAudits.push({
        category: auditName,
        actual: actualResult,
        expected: expectedResult,
        equal: !Object.keys(diff).length,
        diff: diff
      });
    });

    return {
      finalUrl: {
        category: FINAL_URL,
        actual: actual.url,
        expected: expected.url,
        equal: actual.url === expected.url
      },
      audits: collatedAudits
    };
  }
}

const OPERAND_EXPECTATION_REGEXP = /^(<=?|>=?)/;

interface DifferenceInterface {
  getDiff: () => Diff;
  matchesExpectation: () => boolean;
}

class ObjectDifference implements DifferenceInterface {
  private path: string;
  private actual: ActualAudit;
  private expected: ExpectedAudit;

  /**
   * Constructor
   * @param {string} path
   * @param {ActualAudit} actual
   * @param {ExpectedAudit} expected
   */
  constructor(path: string, actual: ActualAudit, expected: ExpectedAudit) {
    this.path = path;
    this.actual = actual;
    this.expected = expected;
  }
  /**
   * Walk down expected result, comparing to actual result. If a difference is found,
   * the path to the difference is returned, along with the expected primitive value
   * and the value actually found at that location. If no difference is found, returns
   * null.
   *
   * Only checks own enumerable properties, not object prototypes, and will loop
   * until the stack is exhausted, so works best with simple objects (e.g. parsed JSON).
   *
   * @return {Diff}
   */
  getDiff(): Diff {
    let diff: Diff = {};

    if (this.matchesExpectation()) return diff;

    // We only care that all expected's own properties are on actual (and not the other way around).
    for (const key of Object.keys(this.expected)) {
      // Bracket numbers, but property names requiring quotes will still be unquoted.
      const keyAccessor = /^\d+$/.test(key) ? `[${key}]` : `.${key}`;
      const keyPath = this.path + keyAccessor;
      const expectedValue = this.expected[key];

      if (!(key in this.actual)) {
        return {path: keyPath, actual: undefined, expected: expectedValue};
      }

      const actualValue = this.actual[key];
      let difference;
      //@todo use factory. P.S. generics should solve this problem
      if (typeof actualValue === 'boolean' && typeof expectedValue === 'boolean') {
        difference = new BooleanDifference(keyPath, { score: actualValue }, { score: expectedValue });
      } else {
        difference = new NoneObjectDifference(keyPath, { score: actualValue }, { score: expectedValue });
      }
      const subDifference = difference.getDiff();
      if (subDifference)
        return diff = subDifference;
    }
    return diff;
  }

  /**
   * Checks if the actual value matches the expectation. Does not recursively search. This supports
   *    - Greater than/less than operators, e.g. "<100", ">90"
   *    - Regular expressions
   *    - Strict equality
   *
   * @return {boolean}
   */
  matchesExpectation(): boolean {
    return Object.is(this.actual, this.expected);
  }
}


class NoneObjectDifference implements DifferenceInterface {
  private path: string;
  private actual: NoneObjectActualAudit;
  private expected: NoneObjectExpectedAudit;

  /**
   * Constructor
   * @param {string} path
   * @param {ActualAudit} actual
   * @param {ExpectedAudit} expected
   */
  constructor(path: string, actual: NoneObjectActualAudit, expected: NoneObjectExpectedAudit) {
    this.path = path;
    this.actual = actual;
    this.expected = expected;
  }

  /**
   * Walk down expected result, comparing to actual result. If a difference is found,
   * the path to the difference is returned, along with the expected primitive value
   * and the value actually found at that location. If no difference is found, returns
   * null.
   *
   * Only checks own enumerable properties, not object prototypes, and will loop
   * until the stack is exhausted, so works best with simple objects (e.g. parsed JSON).
   *
   * @return {Diff}
   */
  getDiff(): Diff {
    let diff: Diff = {};

    if (this.matchesExpectation()) return diff;

    for (const expectationType of Object.keys(this.expected)) {
      const expectedByType = this.expected[expectationType];
      // If they aren't both an object we can't recurse further, so this is the difference.
      if (this.actual.score === null || expectedByType === null || typeof this.actual.score !== 'object' ||
        typeof expectedByType !== 'object' || expectedByType instanceof RegExp) {
        diff = {
          path: this.path,
          actual: this.actual.score,
          expected: this.expected.score
        };
      }
    }
    return diff;
  }

  /**
   * Checks if the actual value matches the expectation. Does not recursively search. This supports
   *    - Greater than/less than operators, e.g. "<100", ">90"
   *
   * @return {boolean}
   */
  matchesExpectation(): boolean {
    const actualValue = this.normalize(this.actual.score);
    const normalizedExpected: NormalizedExpectedScore = {
      warn: this.normalize(this.expected.score.warn),
      error: this.normalize(this.expected.score.error)
    };
    return this.inRange(actualValue, normalizedExpected);
  }

  /**
   * Checks if the actual value in warning and error range
   *
   * @param {number} actual
   * @param {NormalizedExpectedScore} expected
   * @return {boolean}
   */
  private inRange(actual: number, expected: NormalizedExpectedScore): boolean {
    return actual >= expected.error || actual >= expected.warn && actual < expected.error;
  }

  /**
   * Normalizes score value
   *
   * @param {string|number} value
   * @return {number}
   */
  private normalize(value: string|number): number {
    if (typeof value === 'string') {
      return parseInt(value.replace(OPERAND_EXPECTATION_REGEXP, ''));
    } else {
      return value;
    }
  }
}

class BooleanDifference implements DifferenceInterface {
  private path: string;
  private actual: BooleanActualAudit;
  private expected: BooleanExpectedAudit;

  /**
   * Constructor
   * @param {string} path
   * @param {ActualAudit} actual
   * @param {ExpectedAudit} expected
   */
  constructor(path: string, actual: BooleanActualAudit, expected: BooleanExpectedAudit) {
    this.path = path;
    this.actual = actual;
    this.expected = expected;
  }

  /**
   * Walk down expected result, comparing to actual result. If a difference is found,
   * the path to the difference is returned, along with the expected primitive value
   * and the value actually found at that location. If no difference is found, returns
   * null.
   *
   * Only checks own enumerable properties, not object prototypes, and will loop
   * until the stack is exhausted, so works best with simple objects (e.g. parsed JSON).
   * @return {Diff}
   */
  getDiff(): Diff {
    if (this.matchesExpectation()) {
      return {};
    } else {
      return {
        path: this.path,
        actual: this.actual.score,
        expected: this.expected.score
      };
    }
  }

  /**
   * Checks if the actual value matches the expectation. Does not recursively search. This supports booleans
   *
   * @return {boolean}
   */
  matchesExpectation(): boolean {
    return this.actual.score == this.expected.score;
  }
}


class DifferenceFactory {
  /**
   * Find difference comparing to actual result.
   *
   * @param {string} path
   * @param {ActualAudit|*} actual
   * @param {ExpectedAudit|*} expected
   * @return {Diff}
   */
  static findDifference(path: string, actual: any, expected: any): Diff {
    let difference;
    //@todo use generics
    if (actual && typeof actual === 'object') {
      difference = new ObjectDifference(path, actual, expected);
    } else if (actual && typeof actual.score === 'boolean') {
      difference = new BooleanDifference(path, actual, expected);
    } else {
      difference = new NoneObjectDifference(path, actual, expected);
    }
    // @todo add regexp diff class
    return difference.getDiff();
  }
}
