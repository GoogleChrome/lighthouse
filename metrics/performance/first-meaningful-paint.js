/**
 * @license
 * Copyright 2016 Google Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
'use strict';

const DevtoolsTimelineModel = require('devtools-timeline-model');

class FirstMeaningfulPaint {
  /**
   * @param {!Array<!Object>} traceData
   */
  static parse(traceData) {
    const model = new DevtoolsTimelineModel(traceData);
    const events = model.timelineModel().mainThreadEvents();

    // Identify the frameID of the main frame
    const startedInPage = model.tracingModel().devToolsMetadataEvents()
      .filter(e => e.name === 'TracingStartedInPage').slice(-1);
    const frameID = startedInPage[0].args.data.page;

    // Find the start of navigation and our meaningful paint
    const userTiming = events
      .filter(e => e.categoriesString.includes('blink.user_timing'))
      .sort((a, b) => b.ts - a.ts);
    const navStart = userTiming.filter(e => {
      return e.name === 'navigationStart' && e.args.frame === frameID;
    }).slice(-1);
    const conPaint = userTiming.filter(e => {
      return e.name === 'firstContentfulPaint' && e.args.frame === frameID;
    }).slice(-1);

    // report the raw numbers
    if (conPaint.length && navStart.length) {
      return Promise.resolve({
        navigationStart: navStart[0].startTime,
        firstMeaningfulPaint: conPaint[0].startTime
      });
    }
    return Promise.resolve(new Error('First meaningful paint metric not found'));
  }
}

module.exports = FirstMeaningfulPaint;
