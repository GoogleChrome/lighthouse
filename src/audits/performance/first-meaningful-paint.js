/**
 * @license
 * Copyright 2016 Google Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

'use strict';

const FMPMetric = require('../../metrics/first-meaningful-paint');
const Audit = require('../audit');
const TracingProcessor = require('../../lib/traces/tracing-processor');

// Parameters for log-normal CDF scoring. To see the curve:
// https://www.desmos.com/calculator/r7t7qfaaih
const SCORE_LOCATION = Math.log(4000);
const SCORE_SHAPE = 0.5;

class FirstMeaningfulPaint extends Audit {
  /**
   * @override
   */
  static get category() {
    return 'Performance';
  }

  /**
   * @override
   */
  static get name() {
    return 'first-meaningful-paint';
  }

  /**
   * @override
   */
  static get description() {
    return 'First meaningful paint';
  }

  /**
   * @override
   */
  static get optimalValue() {
    return '1,000ms';
  }

  /**
   * Audits the page to give a score for First Meaningful Paint.
   * @see https://github.com/GoogleChrome/lighthouse/issues/26
   * @see https://docs.google.com/document/d/1BR94tJdZLsin5poeet0XoTW60M0SjvOJQttKT-JK8HI/view
   * @param {!Artifacts} artifacts The artifacts from the gather phase.
   * @return {!AuditResult} The score from the audit, ranging from 0-100.
   */
  static audit(artifacts) {
    const traceData = artifacts.traceContents;
    return FMPMetric.parse(traceData).then(data => {
      // there are a few candidates for fMP:
      // * firstContentfulPaint: the first time that text or image content was painted.
      // * fMP basic: paint after most significant layout
      // * fMP page height: basic + scaling sigificance to page height
      // * fMP webfont: basic + waiting for in-flight webfonts to paint
      // * fMP full: considerig both page height + webfont heuristics

      // We're interested in the last of these
      const lastfMPts = data.fmpCandidates
        .map(e => e.ts)
        .reduce((mx, c) => Math.max(mx, c));

      // First meaningful paint (following most significant layout)
      const firstMeaningfulPaint = (lastfMPts - data.navStart.ts) / 1000;

      // Use the CDF of a log-normal distribution for scoring.
      //   < 1100ms: score≈100
      //   4000ms: score≈50
      //   >= 14000ms: score≈0
      const distribution =
          TracingProcessor.getLogNormalDistribution(SCORE_LOCATION, SCORE_SHAPE);
      let score = 100 * distribution.computeComplementaryPercentile(firstMeaningfulPaint);

      return {
        duration: `${firstMeaningfulPaint.toFixed(2)}ms`,
        score: Math.round(score)
      };
    }).catch(err => {
      // Recover from trace parsing failures.
      return {
        score: -1,
        debugString: err.message
      };
    })
    .then(result => {
      return FirstMeaningfulPaint.generateAuditResult({
        value: result.score,
        rawValue: result.duration,
        debugString: result.debugString,
        optimalValue: this.optimalValue
      });
    });
  }
}

module.exports = FirstMeaningfulPaint;
