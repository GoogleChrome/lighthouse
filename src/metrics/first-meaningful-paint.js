/**
 * @license
 * Copyright 2016 Google Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
'use strict';

const FAILURE_MESSAGE = 'Navigation and first paint timings not found.';

// Time to First Meaningful Paint: a layout-based approach
// https://docs.google.com/document/d/1BR94tJdZLsin5poeet0XoTW60M0SjvOJQttKT-JK8HI/edit
class FMP {

  /**
   * @param {!Array<!Object>} traceData
   */
  static parse(traceData) {
    return new Promise((resolve, reject) => {
      if (!traceData || !Array.isArray(traceData)) {
        return reject(new Error(FAILURE_MESSAGE));
      }

      const evts = this.collectEvents(traceData);

      /* eslint-disable no-multi-spaces  */
      const navStart =      evts.navigationStart;
      const fCP =           evts.firstContentfulPaint;
      const fMPbasic =      this.firstMeaningfulPaint(evts, {});
      const fMPpageheight = this.firstMeaningfulPaint(evts, {pageHeight: true});
      const fMPwebfont =    this.firstMeaningfulPaint(evts, {webFont: true});
      const fMPfull =       this.firstMeaningfulPaint(evts, {pageHeight: true, webFont: true});
      /* eslint-enable no-multi-spaces */

      var results = {
        navStart,
        fmpCandidates: [
          fCP,
          fMPbasic,
          fMPpageheight,
          fMPwebfont,
          fMPfull
        ]
      };
      return resolve(results);
    });
  }

  /**
   * @param {!Array<!Object>} traceData
   */
  static collectEvents(traceData) {
    let mainFrameID;
    let navigationStart;
    let firstContentfulPaint;
    let layouts = new Map();
    let paints = [];

    // const model = new DevtoolsTimelineModel(traceData);
    // const events = model.timelineModel().mainThreadEvents();
    const events = traceData;

    // Parse the trace for our key events
    events.filter(e => {
      return e.cat.includes('blink.user_timing') ||
        e.name === 'FrameView::performLayout' ||
        e.name === 'Paint';
    }).forEach(event => {
      // navigationStart == the network begins fetching the page URL
      if (event.name === 'navigationStart' && !navigationStart) {
        mainFrameID = event.args.frame;
        navigationStart = event;
      }
      // firstContentfulPaint == the first time that text or image content was
      // painted. See src/third_party/WebKit/Source/core/paint/PaintTiming.h
      if (event.name === 'firstContentfulPaint' && event.args.frame === mainFrameID) {
        firstContentfulPaint = event;
      }
      // COMPAT: frame property requires Chrome 52 (r390306)
      // https://codereview.chromium.org/1922823003
      if (event.name === 'FrameView::performLayout' && event.args.counters &&
          event.args.counters.frame === mainFrameID) {
        layouts.set(event, event.args.counters);
      }

      if (event.name === 'Paint' && event.args.data.frame === mainFrameID) {
        paints.push(event);
      }
    });

    return {
      navigationStart,
      firstContentfulPaint,
      layouts,
      paints
    };
  }

  /**
   * @param  {any} evts
   */
  static firstContentfulPaint(evts) {
    return (evts.firstContentfulPaint.ts - evts.navigationStart.ts) / 1000;
  }

  /**
   * @param  {any} evts
   * @param  {any} heuristics
   */
  static firstMeaningfulPaint(evts, heuristics) {
    let mostSignificantLayout;
    let significance = 0;
    let maxSignificanceSoFar = 0;
    let pending = 0;

    evts.layouts.forEach((countersObj, layoutEvent) => {
      const counter = val => countersObj[val];

      function heightRatio() {
        const ratioBefore = counter('contentsHeightBeforeLayout') / counter('visibleHeight');
        const ratioAfter = counter('contentsHeightAfterLayout') / counter('visibleHeight');
        return (max(1, ratioBefore) + max(1, ratioAfter)) / 2;
      }

      if (!counter('host') || counter('visibleHeight') === 0) {
        return;
      }

      const layoutCount = counter('LayoutObjectsThatHadNeverHadLayout') || 0;
      // layout significance = number of layout objects added / max(1, page height / screen height)
      significance = (heuristics.pageHeight) ? (layoutCount / heightRatio()) : layoutCount;

      if (heuristics.webFont && counter('hasBlankText')) {
        pending += significance;
      } else {
        significance += pending;
        pending = 0;
        if (significance > maxSignificanceSoFar) {
          maxSignificanceSoFar = significance;
          mostSignificantLayout = layoutEvent;
        }
      }
    });

    const paintAfterMSLayout = evts.paints.find(e => e.ts > mostSignificantLayout.ts);
    return paintAfterMSLayout;
  }
}

module.exports = FMP;

/**
 * Math.max, but with NaN values removed
 */
function max() {
  const args = [...arguments].filter(val => !isNaN(val));
  return Math.max.apply(Math, args);
}
