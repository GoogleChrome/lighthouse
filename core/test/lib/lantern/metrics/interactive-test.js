/**
 * @license
 * Copyright 2024 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */

import assert from 'assert/strict';

import {Interactive} from '../../../../lib/lantern/metrics/interactive.js';
import {FirstMeaningfulPaint} from '../../../../lib/lantern/metrics/first-meaningful-paint.js';
import {getComputationDataFromFixture} from './metric-test-utils.js';
import {readJson} from '../../../test-utils.js';

const trace = readJson('../../../fixtures/traces/progressive-app-m60.json', import.meta);
const devtoolsLog = readJson('../../../fixtures/traces/progressive-app-m60.devtools.log.json', import.meta);
const iframeTrace = readJson('../../../fixtures/traces/iframe-m79.trace.json', import.meta);
const iframeDevtoolsLog = readJson('../../../fixtures/traces/iframe-m79.devtoolslog.json', import.meta);

describe('Metrics: Lantern TTI', () => {
  it('should compute predicted value', async () => {
    const data = await getComputationDataFromFixture({trace, devtoolsLog});
    const result = await Interactive.compute(data, {
      fmpResult: await FirstMeaningfulPaint.compute(data),
    });

    expect({
      timing: Math.round(result.timing),
      optimistic: Math.round(result.optimisticEstimate.timeInMs),
      pessimistic: Math.round(result.pessimisticEstimate.timeInMs),
    }).toMatchSnapshot();
    assert.equal(result.optimisticEstimate.nodeTimings.size, 20);
    assert.equal(result.pessimisticEstimate.nodeTimings.size, 80);
    assert.ok(result.optimisticGraph, 'should have created optimistic graph');
    assert.ok(result.pessimisticGraph, 'should have created pessimistic graph');
  });

  it('should compute predicted value on iframes with substantial layout', async () => {
    const data = await getComputationDataFromFixture({
      trace: iframeTrace,
      devtoolsLog: iframeDevtoolsLog,
    });
    const result = await Interactive.compute(data, {
      fmpResult: await FirstMeaningfulPaint.compute(data),
    });

    expect({
      timing: Math.round(result.timing),
      optimistic: Math.round(result.optimisticEstimate.timeInMs),
      pessimistic: Math.round(result.pessimisticEstimate.timeInMs),
    }).toMatchSnapshot();
    assert.ok(result.optimisticGraph, 'should have created optimistic graph');
    assert.ok(result.pessimisticGraph, 'should have created pessimistic graph');
  });
});
