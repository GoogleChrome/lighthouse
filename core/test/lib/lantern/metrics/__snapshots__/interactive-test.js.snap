// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Metrics: Lantern TTI should compute predicted value 1`] = `
Object {
  "optimistic": 1033,
  "pessimistic": 1069,
  "timing": 1053,
}
`;

exports[`Metrics: Lantern TTI should compute predicted value on iframes with substantial layout 1`] = `
Object {
  "optimistic": 2372,
  "pessimistic": 2386,
  "timing": 2379,
}
`;
